autogen/agentchat/groupchat.py

```diff
@@ -200,20 +193,22 @@ def nested_agents(self) -> List[Agent]:
 
     def next_agent(self, agent: Agent, agents: Optional[List[Agent]] = None) -> Agent:
         """Return the next agent in the list."""
-        if agents is None:
-            agents = self.agents
+        agents = agents or self.agents
 
-        # What index is the agent? (-1 if not present)
-        idx = self.agent_names.index(agent.name) if agent.name in self.agent_names else -1
+        # Ensure the provided list of agents is a subset of self.agents
+        if not set(agents).issubset(set(self.agents)):
+            raise UndefinedNextAgent()
 
-        # Return the next agent
-        if agents == self.agents:
-            return agents[(idx + 1) % len(agents)]
-        else:
-            offset = idx + 1
-            for i in range(len(self.agents)):
-                if self.agents[(offset + i) % len(self.agents)] in agents:
-                    return self.agents[(offset + i) % len(self.agents)]
+        # Try to find the index of the current agent in the provided list
+        idx = self.agents.index(agent) if agent in self.agents else -1
```

Comment:
We need some context on why `self.agent_names` is used currently instead of `self.agents`. @joshkyh @freedeaths @sonichi

Reply:
Hey sorry Eric, I missed this tag somehow, I was working on #1791 this weekend. I'll have a look.

Reply:
Hey @WaelKarkoub, I appreciate the raise UndefinedNextAgent in L200,  and L211. But I'm not confident in agreeing to the logic change in L202-208. Can you please revert back to the old code just for this bit, for me, both are hard cognitively (no improvement).

Reply:
@WaelKarkoub perhaps we can keep the old code as josh suggested?

Reply:
Sounds good, just reverted those changes!